### PR TITLE
feat: cms prototype

### PR DESCRIPTION
**Pull request by simonwkla**
Changes:

- Added a login page and cms landing page
- Added a new `Input` component which extends the Chakra Input Component and applies the color scheme
- Implemented Layout management via a `getLayout` function that every page can use to define what layout it uses (https://nextjs.org/docs/basic-features/layouts#per-page-layouts)
- The `preventTimingAttacks1234` password is now hashed before comparing --> argon throws an error otherwise
- New query field `currentUser` --> returns the currently logged in user
- updated the color logic. By using a custom color scheme (`brand`) we can change the color scheme whenever we want and are not dependant on chakra colors (at the moment brand is using chakra colors though)

### CODE REVIEW

File: app/layout/CmsLayout.tsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+import { useRouter } from "next/router";
+import { FiHome, FiFileText, FiUsers, FiImage } from "react-icons/fi";
+import { Link } from "$app/links";
+
+interface NavElementProps {
+	href: string;
+	name: string;
+	selected: boolean;
+	icon: IconType;
+}
```
Comment (3x071c) [inline app/layout/CmsLayout.tsx:29]: This can be inlined

File: app/layout/CmsLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+		currentUser {
+			firstname
+			lastname
+		}
+	}
+`;
+
+interface CurrentUser {
+	currentUser: Pick<User, "firstname" | "lastname">;
+}
```
Comment (3x071c) [inline app/layout/CmsLayout.tsx:87]: We have Apollo codegen for that :D

File: app/input/Input.tsx [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import {
+	Input as ChakraInput,
+	InputProps,
+	useColorModeValue,
+} from "@chakra-ui/react";
+
+export default function Input(props: PropsWithChildren<InputProps>) {
+	const focusBorderColor = useColorModeValue("brand.500", "brand.300");
+
+	return <ChakraInput {...props} focusBorderColor={focusBorderColor} />;
```
Comment (3x071c) [inline app/input/Input.tsx:11]: This seems oddly specific... why would we change the default `focusBorderColor`?

File: app/layout/CmsLayout.tsx [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+					fontSize={18}
+					textDecoration="none"
+					fontWeight={600}
+					color={color}>
+					{name}
+				</Text>
+			</Flex>
+		</Link>
+	);
+}
```
Comment (3x071c) [inline app/layout/CmsLayout.tsx:60]: This looks large enough to put it in a separate file (in general, it should be one component per file anyway)

File: app/layout/CmsLayout.tsx [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+type CurrentPage = "pages" | "users" | "home" | "files";
+
+interface Props {
+	currentPage: CurrentPage;
+}
```
Comment (3x071c) [inline app/layout/CmsLayout.tsx:66]: Why not inline with the props definition?

File: app/layout/CmsLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	const { loading: userLoading, data: userData } =
+		useQuery<CurrentUser>(CURRENT_USER);
+
+	const router = useRouter();
+
+	const { colorMode, toggleColorMode } = useColorMode();
+
+	if (logoutData) {
+		if (typeof window !== "undefined") {
+			void router.push("/login");
```
Comment (3x071c) [inline app/layout/CmsLayout.tsx:101]: Absolute, untyped paths like that are dangerous. If the CMS ends up at `/cms`, this redirect won't work :(

File: app/layout/CmsLayout.tsx [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+			<Box flexGrow={1}>{children}</Box>
+		</Flex>
+	);
+}
+
+CmsLayout.getLayout = (currentPage: CurrentPage) => {
+	return function CmsLayoutWrapper(page: ReactElement) {
+		return <CmsLayout currentPage={currentPage}>{page}</CmsLayout>;
+	};
+};
```
Comment (3x071c) [inline app/layout/CmsLayout.tsx:189]: Giving CMS pages the ability to modify their own Next.js layout is going to be tricky, I would stick to leaving everything other than content to the developers

File: app/theme.ts [hunk 8]
```diff
@@ -2,16 +2,22 @@ import {
 	extendTheme,
 	withDefaultColorScheme,
 	ThemeConfig,
+	theme,
 } from "@chakra-ui/react";
 
 const config: ThemeConfig = {
-	initialColorMode: "system",
+	initialColorMode: "light",
```
Comment (3x071c) [inline app/theme.ts:9]: `initialColorMode` should account for user preference

File: app/theme.ts [hunk 9]
```diff
@@ -2,16 +2,22 @@ import {
...
 };
 
 export default extendTheme(
 	{ config },
+	{
+		colors: {
+			brand: theme.colors.purple,
```
Comment (3x071c) [inline app/theme.ts:16]: You sure we want to start customizing the color scheme this early? IMO it's quite fine as-is.

File: graphql/schema/User.ts [hunk 10]
```diff
@@ -89,10 +99,14 @@ export const LoginMutation = mutationField("login", {
...
 
+		const PREVENT_TIMING_ATTACKS = await hashPassword(
+			"preventTimingAttacks1234",
+		);
+
 		if (
 			!(await verifyPassword(
 				password,
-				user?.password ?? "preventTimingAttacks123",
+				user?.password ?? PREVENT_TIMING_ATTACKS,
```
Comment (3x071c) [inline graphql/schema/User.ts:109]: This is a critical vulnerability! It allows the (publicized) password "preventTimingAttacks123" to be used to impersonate anyone.

File: pages/login.tsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+	const router = useRouter();
+
+	function handleSubmit() {
+		void login({
+			variables: {
+				password,
+				username,
+			},
+		});
```
Comment (3x071c) [inline pages/login.tsx:38]: Form submissions should be handled with `react-hook-form` to guard user input

File: types/auth.d.ts [hunk 12]
```diff
@@ -1,7 +1,9 @@
 import type { User } from "@prisma/client";
 
+type IronSessionUser = Pick<User, "id">;
+
 declare module "iron-session" {
 	interface IronSessionData {
-		user?: Pick<User, "id">;
+		user?: IronSessionUser;
```
Comment (3x071c) [inline types/auth.d.ts:7]: Why not keep this as a one-liner?

**Review by 3x071c — CHANGES_REQUESTED** [review]

File: app/layout/CmsLayout.tsx (hunk 6 above)
Comment (simonwkla) [reply app/layout/CmsLayout.tsx:101]: What do would you suggest to prevent that from happending?

File: app/theme.ts (hunk 9 above)
Comment (simonwkla) [reply app/theme.ts:16]: I changed it to purple for testing, should have changed it back to blue, my bad ;-;

File: app/layout/CmsLayout.tsx (hunk 5 above)
Comment (simonwkla) [reply app/layout/CmsLayout.tsx:66]: The CurrentPage type is used as a paremter type for the `CmsLayout.getLayout` function

**Comment by 3x071c** [comment]
This doesn't look that mergeable anymore now that we've migrated to Remix. I'll try to re-implement your changes here, but you can have a shot at it first if you want to. I'll close this for now.

**Pull request closed without merging**